### ISSUE TITLE
[Question] Miscellaneous questions

### ISSUE THREAD

**Issue by suargi** [DLR-RM/stable-baselines3#1650]
### ❓ Question

Hi,

I have several questions:

# Entropy scheduler
In SB3 it is possible to define the weight for the entropy loss function that it is used for example in A2C. I would like to define an scheduler for this weight, going for example from 1 at the beginning of the training to 0.01 at the end. As far as I have checked, SB3 uses a constant factor. Would be possible to define an scheduler, maybe using some wrapper/callback?

# A2C continuous policy
In A2C with a continuous policy we need the standard deviation $\sigma$ or the variance $\sigma^2$. What is the output in SB3?

# Vectorized environments
## Monitoring
In a multiple environments setup, each environment is independent from the rest.
For creating the vectorized environment I use:
```
env = make_vec_env("LunarLander-v2",
                   n_envs=N_ENVS,
                   seed=0,
                   monitor_dir=LOG_DIR,
                   env_kwargs={'continuous':True,
                               'enable_wind': False})
```
This created a .csv file for each environment. However, the following code,
```
env = make_vec_env("LunarLander-v2",
                   n_envs=N_ENVS,
                   seed=0,
#                   monitor_dir=LOG_DIR,
                   env_kwargs={'continuous':True,
                               'enable_wind': False})

 env = VecMonitor(venv=env, filename=LOG_DIR)
```
will create a single .csv file. How can I interpret the content of this .csv file?

## Communication
The procedure to communicate data in SB3 is: each environment gathers data and then these observations: state, reward, action and done, are send to the policy for training, i.e., computing gradients.
Is it possible to use the following approach?
Each environment computes the gradients with their data. Then, these gradients can be summed together in the main policy to performer an optimizer update.

## Data gathering implementation
I am a little bit confused on how the data gathering is implemented in a vectorized environment.
We have a policy and N environments.
For every update of the policy:
   1. First interact with the N environments. Each environment runs till the maximum number of iterations per episode or till the terminal state is reached.
   2. When using `A2C` with `n_steps`=10, we collect 10 * N samples from the environments and use them for training.
   3. Update the policy
   4. Further required steps.

I would like to use as much data as possible, eg. collect all steps from all environments. If I set `n_steps` to a very large number, what happens? Will the code use all steps or maybe since the data available is smaller than the specified it will skip that update?

# StopTrainingOnNoModelImprovement callback
I have defined the following code, although the code is not really relevant for my question:
```
stop_train_callback = StopTrainingOnNoModelImprovement(max_no_improvement_evals=3,
                                                           min_evals=5,
                                                           verbose=1)
eval_callback = EvalCallback(env,
                             callback_after_eval=stop_train_callback,
                             best_model_save_path=CHECKPOINT_DIR,
                             log_path=LOG_DIR,
                             eval_freq=max(1000 // N_ENVS, 1),
                             deterministic=True,
                             render=False,
                             verbose=1)
```
I guess we use a single environment for the model evaluation during training.
The reported information is, for example:
```
Eval num_timesteps=3000, episode_reward=-203.71 +/- 82.52 
Episode length: 106.40 +/- 50.03
```
Since we do not specify the number of steps for the evaluation, I guess we evaluate the model for just one episode - till the maximum number of iterations is reached or the episode ends. Therefore, why is the mean episode length a float number (not an integer) and the standard deviation different from zero?

# Custom callback variables
In a custom callback: ```class MyCustomCallback(BaseCallback):``` we have access tot he variable `self.locals`. Inside locals, there is the array called `rewards`. The latter corresponds to the mean, sum or last rewards of the batch? For instance:
```
class MyCustomCallback(BaseCallback):
        def __init__(self, verbose=0, solved_threshold=200):
            super(MyCustomCallback, self).__init__(verbose)
            # Do things

        def _on_step(self) -> bool:
            return True

        def _on_rollout_end(self) -> None:
            print(self.locals.get('rewards'))
```

I hope everything is clear. Let me know if it is not the case.

Thanks!

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] If code there is, it is [minimal and working](https://github.com/DLR-RM/stable-baselines3/issues/982#issuecomment-1197044014)
- [X] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by araffin** [comment]
### Entropy scheduler

> Would be possible to define an scheduler, maybe using some wrapper/callback?

yes, duplicate of https://github.com/DLR-RM/stable-baselines3/issues/316 and https://github.com/DLR-RM/stable-baselines3/issues/628

### A2C continuous policy

not sure what you mean. This kind of detail should be covered by the ppo blog post (see link in our doc).

### Monitoring

> How can I interpret the content of this .csv file?

In case of doubt, best is to take a look at the code.
It can be interpreted as a concatenation of the multiple csv files while preserving the correct order (chronological order).

### Communication

Currently no. You would probably need MPI: https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/issues/11

### Data gathering implementation

probably also covered by the ppo blog post and by spinning up tutorial (see link in doc).
A2C is very close to REINFORCE/vanilla policy gradient in how it uses the data.


### StopTrainingOnNoModelImprovement

>  I guess we evaluate the model for just one episode - till the maximum number of iterations is reached or the episode ends. 

default is 5 eval episodes if I recall (see doc/code of `EvalCallback`).


### Custom callback variables

> The latter corresponds to the mean, sum or last rewards of the batch? 

should correspond to the immediate reward

**Comment by suargi** [comment]
# A2C continuous policy
> not sure what you mean. This kind of detail should be covered by the ppo blog post (see link in our doc).

According to the literature, the actuator of the A2C outputs the mean value of the PDF and the critic outputs the standard deviation, or the variance.

I traced the code but could not figure it out.

**Comment by araffin** [comment]
> According to the literature, the actuator of the A2C outputs the mean value of the PDF and the critic outputs the standard deviation, or the variance.

There seems to be a misunderstanding, the actor outputs mean and log std of the Gaussian distribution.
The critic has nothing to do with action selection.

The relevant pieces of code are:
https://github.com/DLR-RM/stable-baselines3/blob/f4ec0f6afa1a23b0e0b746174cd0074471cc0b89/stable_baselines3/common/distributions.py#L148-L150

and

https://github.com/DLR-RM/stable-baselines3/blob/f4ec0f6afa1a23b0e0b746174cd0074471cc0b89/stable_baselines3/common/policies.py#L569-L571

and

https://github.com/DLR-RM/stable-baselines3/blob/f4ec0f6afa1a23b0e0b746174cd0074471cc0b89/stable_baselines3/common/policies.py#L646

**Comment by suargi** [comment]
Thank you. Everything is now clear
